system/src/ledger/ledger_manager.cpp

```diff
@@ -0,0 +1,1599 @@
+        }
+        runNext(retryTime_ - now);
+        return 0;
+    }
+    if (state_ != State::READY || (!pendingState_ && !resubscribe_)) {
+        // Some task is in progress, the manager is not in an appropriate state, or there's nothing to do
+        return 0;
+    }
+    if (pendingState_ & PendingState::GET_INFO) {
+        LOG(INFO, "Requesting ledger info");
+        CHECK(sendGetInfoRequest());
+        return 1;
+    }
+    if ((pendingState_ & PendingState::SUBSCRIBE) || resubscribe_) {
+        LOG(INFO, "Subscribing to ledger updates");
+        CHECK(sendSubscribeRequest());
+        return 1;
+    }
+    unsigned nextSyncDelay = 0;
+    if (pendingState_ & PendingState::SYNC_TO_CLOUD) {
+        if (now >= nextSyncTime_) {
+            uint64_t t = 0;
+            LedgerSyncContext* ctx = nullptr;
+            for (auto& c: contexts_) {
+                if (c->pendingState & PendingState::SYNC_TO_CLOUD) {
+                    if (now >= c->syncTime) {
+                        ctx = c.get();
+                        break;
+                    }
+                    if (!t || c->syncTime < t) {
+                        t = c->syncTime;
+                    }
+                }
+            }
+            if (ctx) {
+                LOG(TRACE, "Synchronizing ledger: %s", ctx->name);
+                CHECK(sendSetDataRequest(ctx));
+                return 1;
+            }
+            // Timestamp in nextSyncTime_ was outdated
+            nextSyncTime_ = t;
+        }
+        nextSyncDelay = nextSyncTime_ - now;
+    }
+    if (pendingState_ & PendingState::SYNC_FROM_CLOUD) {
+        LedgerSyncContext* ctx = nullptr;
+        for (auto& c: contexts_) {
+            if (c->pendingState & PendingState::SYNC_FROM_CLOUD) {
+                ctx = c.get();
+                break;
+            }
+        }
+        if (ctx) {
+            LOG(TRACE, "Synchronizing ledger: %s", ctx->name);
+            CHECK(sendGetDataRequest(ctx));
+            return 1;
+        }
+    }
+    if (nextSyncDelay > 0) {
+        runNext(nextSyncDelay);
+    }
+    return 0;
+}
+
+int LedgerManager::notifyConnected() {
+    if (state_ != State::OFFLINE) {
+        return SYSTEM_ERROR_INVALID_STATE;
+    }
+    LOG(TRACE, "Connected");
+    startSync();
+    return 0;
+}
+
+void LedgerManager::notifyDisconnected(int /* error */) {
+    if (state_ == State::OFFLINE) {
+        return;
+    }
+    if (state_ == State::NEW) {
+        LOG_DEBUG(ERROR, "Unexpected manager state: %d", (int)state_);
+        return;
+    }
+    LOG(TRACE, "Disconnected");
+    reset();
+    retryDelay_ = 0;
+    state_ = State::OFFLINE;
+}
+
+int LedgerManager::receiveRequest(CoapMessagePtr& msg, int reqId) {
+    if (state_ < State::READY) {
+        return SYSTEM_ERROR_INVALID_STATE;
+    }
+    // Get the request type. XXX: It's assumed that the message fields are encoded in order of their
+    // field numbers, which is not guaranteed by the Protobuf spec in general
```

Comment:
Can you add a test in https://github.com/particle-iot/device-os-protobuf/blob/main/src/index.test.js to make sure this remains true?

Reply:
That's a good idea. There are integration tests in this PR that validate this indirectly but we only run them in production environment.

Reply:
Added a test in device-os-protobuf: https://github.com/particle-iot/device-os-protobuf/pull/29.